TACC/tup-ui
# Add plugin for rapid generation of staff pages - update UI & UX

## Overview

Improve UI and UX and data.

## Related

- [TUP-548](https://jira.tacc.utexas.edu/browse/TUP-548)
- builds atop #253
- includes:
    - #294
    - #295

## Changes

- **added** admin interface for plugin instances
- **changed** plugin name to be translatable
- **added** field help text and custom labels
- **changed** which fields are required
- **added** form layout and field organization
- **added** `post_nomials` and `title2`
- **changed** `photo` to not be required
- **removed** `experience`[^1]
- **added** styles for new ui requests
- **added** image fallback for missing photo
- **added** script for new ui request
- **changed** markup to match live pages
- **added** snippet for ui and ux on manual pages

## Testing

<details><summary>Setup</summary>

0. Checkout and pull the branch of this PR.
1. Run the CMS locally, and be a superuser.
	<sup>After [TACC/tup-ui:README.md](https://github.com/TACC/tup-ui/blob/v1.0.10/README.md#tacc) step 7, continue with [TACC/Core-CMS/:README.md "Start Project"](https://github.com/TACC/Core-CMS/blob/v3.12.0-alpha.10/README.md#start-project) step 2.</sup>
2. `docker exec -it tup_cms /bin/bash`
    then `python manage.py migrate staff_profile`
    <sup>Unless you already ran `python manage.py migrate` cuz you set up fresh local env on this branch.</sup>
3. Add or edit a "Staff Profile" plugin instance.

</details>

1. <details><summary>Verify job title and dept. combinations render as expected.</summary>

	- one title and no dept.:
    	> Title
	- one title and a dept.:
    	> **Title**
    	> Dept
	- two titles and a dept.:
    	> **Title, Dept.**
    	> **Title2**
	- two titles and no dept.:
    	> **Title**
    	> **Title2**

	</details>

2. <details><summary>Verify image "alt text" possibilities:</summary>

    1. with "Alt. text" (add via Filer), uses that for `alt` attr.
    2. with no alt text (via Filer), uses full name for `alt` attr.

	</details>

3. <details><summary>Verify fields that are not required, when not provided, do not render excess markup.</summary>

    1. empty "Phone" should not render "Phone: | "
    2. empty "Group / Department" should not render ", " nor `<br /><small></small>`
    3. empty "Other" content should not render `<div content-block">…</div>`

	</details>

4. <details><summary>Verify required fields are only names, title, and e-mail.</summary>

    Any field required that is not filled will render error upon form submission.

	</details>

5. Verify a missing photo will render the UT Seal.

6. <details><summary>Verify a missing photo will render the UT Seal.</summary>

    Any required field left empty will cause error to render after form submission.

	</details>

7. <details><summary>Verify design is responsive in specific ways:</summary>

    - On narrow screen (<768px):
    	- Photo is beneath name and title and dept.
    	- Education is at bottom of page
    	- Education lacks green border beside heading
    	- Education lacks extra padding
    - On wide screen (≥768px):
    	- Photo is above Education, right of name and title and dept.
    	- Education is below Photo, right of name and title and dept.
    	- Education has green border beside heading
    	- Education has extra padding

	</details>

8. <details><summary>Verify staff profiles can be browsed and edited via admin interface.</summary>

    Either, open http://localhost:8000/admin/staff_profile/staffprofileplugin.

	Or, from http://localhost:8000/admin/, scroll to "TACC Staff Profiles" and click "Staff profile plugins".

	</details>

## UI

1. <details><summary>Job Title & Dept. Differences</summary>

	| Page | Form |
	| - | - |
	| ![ian page](https://github.com/TACC/tup-ui/assets/62723358/8b205b72-3abc-4bbe-9f58-6afcc706962b) | ![ian form](https://github.com/TACC/tup-ui/assets/62723358/4b46bc04-0cae-44f7-8159-eb8317470b85) |
	| ![stephen page](https://github.com/TACC/tup-ui/assets/62723358/f054c80d-70a4-4c26-9d7c-4cf403937610) | ![stephen form](https://github.com/TACC/tup-ui/assets/62723358/7fe0c74d-b716-455a-bb67-9423ad17d9cf) |
	| ![dan page](https://github.com/TACC/tup-ui/assets/62723358/9cb7ab31-2d2b-439c-8aee-e9f99f1bebdc) | ![dan form](https://github.com/TACC/tup-ui/assets/62723358/e61e65c3-5a30-40eb-8304-78228ad5ada3) |

	</details>
<br />

5. <details><summary>and 6. Photo Not Required & Fallback</summary>

	https://github.com/TACC/tup-ui/assets/62723358/ed0ea358-4e84-4441-9b9f-632130cbb9dd

	</details>
<br />

7. <details><summary>Responsive Design</summary>

	| Wide | Narrow |
	| - | - |
	| ![staff profile - wide](https://github.com/TACC/tup-ui/assets/62723358/58a4e4da-344b-42db-88ea-26a0fe5c385e) | ![staff profile - narrow](https://github.com/TACC/tup-ui/assets/62723358/a71cb44e-b21f-4707-9e1e-3f4af021b219) |

	</details>

8. <details><summary>Admin Interface</summary>

	| Admin | Table | Form |
	| - | - | - |
	| ![1 admin list](https://github.com/TACC/tup-ui/assets/62723358/d93a40a8-7d07-4a25-aa68-1235c188fb93) | ![2 staff_profile table](https://github.com/TACC/tup-ui/assets/62723358/39bff81d-4d20-48b2-8b34-ee3ab65ccef6) | ![3 staff_profile form](https://github.com/TACC/tup-ui/assets/62723358/06a3983c-e936-44bc-b22f-d95840af10c5) |

	</details>

## Notes

[CMD Proposals - Staff Profile Layout & Data](https://confluence.tacc.utexas.edu/pages/viewpage.action?pageId=311558206)

[^1]: H.P. says Professional Experience is for leadership, as wanted; rare in general; and easily appended to staff Bio.
